doc/source/data/performance-tips.rst

```diff
@@ -25,17 +25,122 @@ Tuning read parallelism
+For example, the following code will batch multiple files into the same read task to avoid creating blocks that are too large.
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Repeat the iris.csv file 16 times.
+    ds = ray.data.read_images(["example://iris.csv"] * 16)
+    print(ds.materialize())
+    # 2023-11-20 14:28:47,597 INFO plan.py:760 -- Using autodetected parallelism=4 for stage ReadCSV to satisfy parallelism at least twice the available number of CPUs (2).
+    # MaterializedDataset(
+    #    num_blocks=4,
+    #    num_rows=2400,
+    #    ...
+    # )
+
+But suppose that you knew that you wanted to read all 16 files in parallel.
+This could, for example, because you know that additional CPUs should get added to the cluster by the autoscaler or because you want the downstream stage to transform each file's contents in parallel.
+You can get this behavior by setting the ``parallelism`` parameter.
+Notice how the number of output blocks is equal to ``parallelism`` in the following code:
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Repeat the iris.csv file 16 times.
+    ds = ray.data.read_images(["example://iris.csv"] * 16, parallelism=16)
+    print(ds.materialize())
+    # MaterializedDataset(
+    #    num_blocks=16,
+    #    num_rows=2400,
+    #    ...
+    # )
+
+
+When using the default autodetected ``parallelism``, Ray Data attempts to cap each task's output to :class:`DataContext.target_max_block_size <ray.data.context.DataContext>` many bytes.
+Note however that Ray Data cannot perfectly predict the size of each task's output, so it is possible that each task produces one or more output blocks.
+Thus, the total blocks in the final :class:`~ray.data.Dataset` may differ from the specified ``parallelism``.
+Here's an example where we manually specify ``parallelism=1``, but the one task will still produce multiple blocks in the materialized Dataset:
+
+.. testcode::
+
+    import ray
+    # Pretend we have two CPUs.
+    ray.init(num_cpus=2)
+
+    # Generate ~400MB of data.
+    ds = ray.data.range_tensor(5_000, shape=(10_000, ), parallelism=1)
+    print(ds.materialize())
+    # MaterializedDataset(
+    #    num_blocks=3,
+    #    num_rows=5000,
+    #    schema={data: numpy.ndarray(shape=(10000,), dtype=int64)}
+    # )
+
+
+Currently, Ray Data can assign at most one read task per input file.
+Thus, if the number of input files is smaller than ``parallelism``, the number of read tasks is capped to the number of input files.
+To ensure that downstream transforms can still execute with the desired parallelism, Ray Data will split the read tasks' outputs into a total of ``parallelism`` blocks and disable fusing with the downstream transform.
+In other words, the read tasks' output blocks are materialized to Ray's object store before any map stage executes.
```

Comment:
this isn't accurate. for `read_parquet`, the file will be read in small batches and data output will still be streamed. 
for `read_csv`, the task will read all contents as a block in worker heap memory, and then chunk and yield smaller blocks to object store. so not all blocks are "materialized in object store".

Reply:
Ah right, good catch. This should say something like "are materialized to Ray's object store before the downstream map task can execute", but yes the execution is still streaming.